### PR TITLE
add share command

### PR DESCRIPTION
**Pull request by 12944qwerty**
I have no idea if this works, so please test

### CODE REVIEW

**Comment by 12944qwerty** [comment]
I'm considering making `--no-embed` the default (it won't show the embed). What do y'all think

File: src/Powercord/coremods/moduleManager/commands/share.js [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+const { openURL } = getModule([ 'openURL' ], false);
+const { ComponentDispatch } = getModule([ 'ComponentDispatch' ], false);
+
+function formatUrl (url) {
+  return url
+    .replace('.git', '')
+    .replace('git@github.com:', 'https://github.com/')
+    .replace('url = ', '');
+}
```
Comment (asportnoy) [inline src/Powercord/coremods/moduleManager/commands/share.js:14]: Let's make a shared util to get git into of a plugin. Can also include things like the branch, but don't worry about that for now. Just add a property for the web URL.

File: src/Powercord/coremods/moduleManager/commands/share.js [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  return {
+    commands: flags.filter(flag => flag.startsWith(lastArg) && !args.includes(flag))
+      .map(x => ({ command: x })),
+    header: 'options list'
+  };
+}
+
+module.exports = {
+  command: 'share',
+  description: 'Share a plugin or theme that you have',
```
Comment (asportnoy) [inline src/Powercord/coremods/moduleManager/commands/share.js:44]:
```suggestion
  description: 'Share an installed plugin or theme',
```

File: src/Powercord/coremods/moduleManager/commands/share.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+      if (args.includes('--no-send')) {
+        appendText(url);
+        return;
+      }
+      return {
+        send: true,
+        result: url
+      };
+    }
+    return resp(false, 'Unable to find a url in the .git config file');
```
Comment (asportnoy) [inline src/Powercord/coremods/moduleManager/commands/share.js:93]: i18n
Make the message less technical

File: src/Powercord/coremods/moduleManager/commands/share.js [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    return {
+      header: 'replugged entities list',
+      commands: [
+        ...plugins.map(plugin => ({
+          command: plugin.entityID,
+          description: `Plugin - ${plugin.manifest.description}`
+        })),
+        ...themes.map(theme => ({
+          command: theme.entityID,
+          description: `Theme - ${theme.manifest.description}`
```
Comment (asportnoy) [inline src/Powercord/coremods/moduleManager/commands/share.js:116]: i18n (may be able to use an existing key)

**Review by asportnoy — CHANGES_REQUESTED** [review]

**Comment by asportnoy** [comment]
> I'm considering making `--no-embed` the default (it won't show the embed). What do y'all think

I think default to install link with no embed. You can then remove the flags for those options.

File: src/Powercord/coremods/moduleManager/commands/share.js (hunk 1 above)
Comment (12944qwerty) [reply src/Powercord/coremods/moduleManager/commands/share.js:14]: What?

Comment (lleyton) [reply src/Powercord/coremods/moduleManager/commands/share.js:14]: I think he means to make it a separate utility function.

Comment (asportnoy) [reply src/Powercord/coremods/moduleManager/commands/share.js:14]: Yeah, create a function in the coremod utils folder that takes a plugin/theme ID import and the type and returns an object with some git-related info. For now, just the web URL is fine, but in the future we can add branch, revision, etc.

File: src/Powercord/coremods/util.js [hunk 5]
```diff
@@ -99,5 +100,33 @@ module.exports = {
     Experiments="${experimentOverrides > 0 ? Object.keys(getExperimentOverrides()).join(', ') : 'n/a'}"
     Labs="${enabledLabs.length ? enabledLabs.map(e => e.name).join(', ') : 'n/a'}"
     Plugins="${plugins.join(', ')}"\`\`\``;
+  },
```
Comment (asportnoy) [inline src/Powercord/coremods/util.js:103]: This should be in the moduleManager util folder. Also make sure you import everything you need.

File: src/Powercord/coremods/util.js [hunk 6]
```diff
@@ -99,5 +100,33 @@ module.exports = {
...
+  
+  formatURL(url) {
+    return url
+      .replace('.git', '')
+      .replace('git@github.com:', 'https://github.com/')
+      .replace('url = ', '');
+  },
+  
+  getURL(entity, type=null) {
```
Comment (asportnoy) [inline src/Powercord/coremods/util.js:112]: Rename this to `getWebURL`

File: src/Powercord/coremods/util.js [hunk 7]
```diff
@@ -99,5 +100,33 @@ module.exports = {
     Experiments="${experimentOverrides > 0 ? Object.keys(getExperimentOverrides()).join(', ') : 'n/a'}"
     Labs="${enabledLabs.length ? enabledLabs.map(e => e.name).join(', ') : 'n/a'}"
     Plugins="${plugins.join(', ')}"\`\`\``;
+  },
+  
+  formatURL(url) {
```
Comment (asportnoy) [inline src/Powercord/coremods/util.js:105]: Rename this to `formatGitURL`

File: src/Powercord/coremods/moduleManager/commands/share.js [hunk 8]
```diff
@@ -41,42 +34,31 @@ function autocompleteFlags (args) {
...
     const isPlugin = args.includes('--plugin') || powercord.pluginManager.plugins.has(id);
     if (!args.includes('--theme') && !args.includes('--plugin')) {
       const isTheme = powercord.styleManager.themes.has(id);
 
       if (!isPlugin && !isTheme) { // No match
-        return resp(false, `Could not find plugin or theme matching "${id}".`);
+        return resp(false, Messages.REPLUGGED_ENTITY_NOT_FOUND.replace('{id}', id);
       } else if (isPlugin && isTheme) { // Duplicate name
-        return resp(false, `"${id}" is in use by both a plugin and theme. You will have to specify with --theme or --plugin flag.`);
+        return resp(false, Messages.REPLUGGED_COMMAND_SHARE_BOTH_ENTITY.replace('{id}', id);
```
Comment (asportnoy) [inline src/Powercord/coremods/moduleManager/commands/share.js:47]:
```suggestion
        return resp(false, Messages.REPLUGGED_COMMAND_SHARE_BOTH_ENTITY.format({ id });
```

File: src/Powercord/coremods/moduleManager/commands/share.js [hunk 9]
```diff
@@ -31,7 +24,7 @@ function autocompleteFlags (args) {
   if (!lastArg.startsWith('-')) {
     return false;
   }
-  const flags = [ '--repo', '--no-repo', '--open', '--no-send', '--embed', '--no-embed', '--theme', '--plugin' ];
+  const flags = [ '--repo', '--open', '--no-send', '--theme', '--plugin' ];
```
Comment (asportnoy) [inline src/Powercord/coremods/moduleManager/commands/share.js:27]:
```suggestion
  const flags = [ '--repo', '--open', '--no-send', '--embed', '--theme', '--plugin' ];
```

File: src/Powercord/coremods/moduleManager/commands/share.js [hunk 10]
```diff
@@ -41,42 +34,31 @@ function autocompleteFlags (args) {
...
       }
-      if ((!args.includes('--no-repo')) || args.includes('--repo')) {
-        if ((args.includes('--no-embed')) && !args.includes('--embed')) {
-          url = `<${url}>`;
-        }
+      if (args.includes('--repo')) {
+        url = `<${url}>`;
       } else {
         url = `<https://replugged.dev/install?url=${url}>`;
       }
```
Comment (asportnoy) [inline src/Powercord/coremods/moduleManager/commands/share.js:64]: re-implement `--embed` flag
```suggestion
      if (!args.includes('--repo')) {
        url = `https://replugged.dev/install?url=${url}`;
      }
      if (!args.includes('--embed')) {
        url = `<${url}>`;
      }
```

File: src/Powercord/coremods/moduleManager/commands/share.js [hunk 11]
```diff
@@ -90,7 +72,7 @@ module.exports = {
         result: url
       };
     }
-    return resp(false, 'Unable to find a url in the .git config file');
+    return resp(false, Messages.REPLUGGED_COMMAND_SHARE_URL_NOT_FOUND.replace('{id}', id);
```
Comment (asportnoy) [inline src/Powercord/coremods/moduleManager/commands/share.js:75]:
```suggestion
    return resp(false, Messages.REPLUGGED_COMMAND_SHARE_URL_NOT_FOUND.format({ id });
```

File: i18n/en-US.json [hunk 12]
```diff
@@ -36,6 +36,9 @@
     "REPLUGGED_COMMAND_RELOAD_DESC": "Reload a plugin/theme",
     "REPLUGGED_COMMAND_RELOAD_RELOADED": "{type} \"{id}\" reloaded!",
     "REPLUGGED_COMMAND_SAY_DESC": "Sends the specified arguments.",
+    "REPLUGGED_COMMAND_SHARE_DESCRIPTION": "Share an installed plugin or theme",
+    "REPLUGGED_COMMAND_SHARE_BOTH_ENTITY": "{id} is in use by both a plugin and theme. You will have to specify with --theme or --plugin flag.",
+    "REPLUGGED_COMMAND_SHARE_URL_NOT_FOUND": "Unable to find a url for the plugin {id}",
```
Comment (asportnoy) [inline i18n/en-US.json:41]:
```suggestion
    "REPLUGGED_COMMAND_SHARE_URL_NOT_FOUND": "Unable to find a URL for {id}",
```

File: src/Powercord/coremods/moduleManager/commands/share.js [hunk 13]
```diff
@@ -42,25 +39,29 @@ module.exports = {
...
+    let url = getWebURL(entity);
+
     if (url !== '') {
       if (args.includes('--open')) {
         openURL(url);
         return;
       }
       if (args.includes('--repo')) {
-        url = `<${url}>`;
+        url = `${url}`;
```
Comment (asportnoy) [inline src/Powercord/coremods/moduleManager/commands/share.js:58]: This is unnecessary, remove it and flip the condition

File: src/Powercord/coremods/moduleManager/util/misc.js [hunk 14]
```diff
@@ -42,3 +45,29 @@ exports.resp = (success, description) => ({
...
+  .replace('.git', '')
+  .replace('git@github.com:', 'https://github.com/')
+  .replace('url = ', '');
+
+exports.getWebURL = (entity, type = null) => {
+  if (typeof entity === 'string') {
+    if (type) {
+      entity = type === 'plugin' ? powercord.pluginManager.get(entity) : powercord.styleManager.get(entity);
+    } else {
+      return false;
```
Comment (asportnoy) [inline src/Powercord/coremods/moduleManager/util/misc.js:59]:
```suggestion
      return null;
```

**Review by asportnoy — CHANGES_REQUESTED** [review]
There are still unresolved issues from my past 2 reviews. Please make sure you're addressing **all** of them and mark them as resolved once you do.

**Review by lleyton — APPROVED** [review]
LGTM

File: src/Powercord/coremods/moduleManager/commands/share.js [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    const plugins = Array.from(powercord.pluginManager.plugins.values())
+      .filter(plugin => plugin.entityID.toLowerCase().includes(id?.toLowerCase()));
+
+    const themes = Array.from(powercord.styleManager.themes.values())
+      .filter(theme => theme.entityID.toLowerCase().includes(id?.toLowerCase()));
+
+    return {
+      header: 'replugged entities list',
```
Comment (asportnoy) [inline src/Powercord/coremods/moduleManager/commands/share.js:88]: Still needs i18n and to be rewritten

**Review by asportnoy — CHANGES_REQUESTED** [review]

**Review by asportnoy — APPROVED** [review]

**Pull request merged**